src/qibojit/custom_operators/raw_kernels.py

```diff
@@ -77,158 +83,249 @@ __device__ void _apply_two_qubit_gate(T& state0, T& state1, T& state2, T& state3
   i = ((long)((long)i >> m2) << (m2 + 1)) + (i & (tk2 - 1));
-  _apply_two_qubit_gate<T>(state[i], state[i + uk1], state[i + uk2], state[i + uk1 + uk2], gate);
+  _apply_two_qubit_gate(state[i], state[i + uk1], state[i + uk2], state[i + uk1 + uk2], gate);
 }
+"""
 
 
+apply_fsim_kernel = f"""
+#include <cupy/complex.cuh>
+{_apply_fsim}"""+"""
 // C++ implementation of gates.py:apply_fsim_kernel()
 // The portion of code before the parallel for of the Python
 // method is in backends.py:CupyBackend.two_qubit_base()
-template<typename T>
+extern "C"
 __global__ void apply_fsim_kernel(T* state, long tk1, long tk2,
                                   int m1, int m2, long uk1, long uk2,
                                   const T* gate) {
   const long g = blockIdx.x * blockDim.x + threadIdx.x;
   long i = ((long)((long)g >> m1) << (m1 + 1)) + (g & (tk1 - 1));
   i = ((long)((long)i >> m2) << (m2 + 1)) + (i & (tk2 - 1));
-  _apply_fsim<T>(state[i + uk1], state[i + uk2], state[i + uk1 + uk2], gate);
+  _apply_fsim(state[i + uk1], state[i + uk2], state[i + uk1 + uk2], gate);
 }
+"""
 
 
+apply_swap_kernel = f"""
+#include <cupy/complex.cuh>
+{_apply_x}"""+"""
 // C++ implementation of gates.py:apply_swap_kernel()
-template<typename T>
+extern "C"
 __global__ void apply_swap_kernel(T* state, long tk1, long tk2,
                                   int m1, int m2, long uk1, long uk2) {
   const long g = blockIdx.x * blockDim.x + threadIdx.x;
   long i = ((long)((long)g >> m1) << (m1 + 1)) + (g & (tk1 - 1));
   i = ((long)((long)i >> m2) << (m2 + 1)) + (i & (tk2 - 1));
-  _apply_x<T>(state[i + tk2], state[i + tk1]);
+  _apply_x(state[i + tk2], state[i + tk1]);
 }
+"""
 
 
+multicontrol_apply_gate_kernel = f"""
+#include <cupy/complex.cuh>
+{_apply_gate}
+{multicontrol_index}"""+"""
 // C++ implementation of gates.py:multicontrol_apply_gate_kernel()
-template<typename T>
+extern "C"
 __global__ void multicontrol_apply_gate_kernel(T* state, long tk, int m, const T* gate,
                                                const int* qubits, int ncontrols) {
   const long g = blockIdx.x * blockDim.x + threadIdx.x;
   const long i = multicontrol_index(qubits, g, ncontrols);
-  _apply_gate<T>(state[i - tk], state[i], gate);
+  _apply_gate(state[i - tk], state[i], gate);
 }
+"""
 
 
+multicontrol_apply_x_kernel = f"""
+#include <cupy/complex.cuh>
+{_apply_x}
+{multicontrol_index}"""+"""
 // C++ implementation of gates.py:multicontrol_apply_x_kernel()
-template<typename T>
+extern "C"
 __global__ void multicontrol_apply_x_kernel(T* state, long tk, int m,
                                             const int* qubits, int ncontrols) {
   const long g = blockIdx.x * blockDim.x + threadIdx.x;
   const long i = multicontrol_index(qubits, g, ncontrols);
-  _apply_x<T>(state[i - tk], state[i]);
+  _apply_x(state[i - tk], state[i]);
 }
+"""
 
 
+multicontrol_apply_y_kernel = f"""
+#include <cupy/complex.cuh>
+{_apply_y}
+{multicontrol_index}"""+"""
```

Comment:
Do we really need the string addition (`+`) here? Wouldn't something like:
```Python
multicontrol_apply_y_kernel = f"""
#include <cupy/complex.cuh>
{_apply_y}
{multicontrol_index}
// C++ implementation of gates.py:multicontrol_apply_y_kernel()
extern "C"
___global___ void ...
"""
```
have the same effect?

Reply:
It's a concatenation between a f-string and a standard one. With your suggestion, the body of the kernel within brackets would be considered as Python code to be evaluated.

Reply:
I see, you are right. Thank you for the explanation and updates.